### DISCUSSION TITLE
How to test in a github workflow?

### DISCUSSION THREAD

**Discussion by DanielT** [PyO3/maturin#1744] [Q&A] [answered]
I have created a python module using pyo3 - [autosar-data-py](https://github.com/DanielT/autosar-data-py)
Next I created some tests for it using pytest.
On my machine, I can set up a virtualenv, run `maturin develop` to build and install my module, and then run `pytest` to test it.

I cannot get this to work in a Github workflow. `maturin develop` runs, but pytest cannot find my module. I think the virtualenv is not getting activated, or not remaining activated after that step.

I have also considered that it might be possible to run `maturin build`, then (somehow) get the filename of the wheel that was built and then run `pip install $(filename)`.
It is not clear to me how I would get the filename.

Does anyone have suggestions, or a working example I could copy/imitate?

**Comment by messense** [comment, marked as answer]
Have you tried `maturin generate-ci github --pytest`?

**Reply by DanielT** [reply]
No, I haven't tried that. In fact, I didn't know it was possible.
I even searched for "maturin generate-ci" and "maturin pytest" before I posted.

Is there some documentation for the generate-ci command that I didn't find, or is it undocumented?

I just realized that you implemented generate-ci.
Thank you!
I was very surprised - in the best possible way - by how well the automatic build of the wheels for all targets worked.
Getting that done by hand would have taken me much longer.
